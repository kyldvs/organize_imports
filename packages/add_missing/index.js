/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

/**
 * Adds missing imports.
 *
 *   type Options = ?{
 *     importKind?: 'import' | 'require' | 'auto',
 *   };
 *
 */
export default function add_missing(_options) {
  const options = _options || {};
  return function babel_plugin({types: t}) {
    return {
      visitor: {
        Program(path) {
          // Cache of requires to add, based on kind.
          const CACHE = {
            value: new Set(),
            type: new Set(),
          };

          // Determine the kind of import we should add by default.
          let importKind = options.importKind || 'auto';
          const extraVisitors = importKind === 'auto'
            ? {
              ImportDeclaration(path) {
                if (path.node.importKind === 'value') {
                  importKind = 'import';
                }
              },
            }
            : {};

          // Traverse the program, then we will add all the requires.
          path.traverse({
            ...extraVisitors,
            Identifier(path) {
              const name = path.node.name;
              if (path.scope.hasBinding(name)) {
                return;
              }
              const kind = t.isGenericTypeAnnotation(path.parentPath.node)
                ? 'type'
                : 'value';
              CACHE[kind].add(name);
            },
          });

          // If it's still auto after the traversal we should use `require`.
          if (importKind === 'auto') {
            importKind = 'require';
          }

          // Add all the requires that we found.
          const added = new Set();
          const importNodes = [];
          CACHE.value.forEach(name => {
            if (!added.has(name)) {
              added.add(name);
              importNodes.push(createImportNode(t, name, importKind));
            }
          });
          CACHE.type.forEach(name => {
            if (!added.has(name)) {
              added.add(name);
              importNodes.push(createImportNode(t, name, 'type'));
            }
          });
          path.node.body = addImportsToBody(t, importNodes, path.node.body);
        },
      },
    };
  };
}

function createImportNode(t, name, kind) {
  if (kind === 'type') {
    const result = t.importDeclaration(
      [t.importDefaultSpecifier(t.identifier(name))],
      t.stringLiteral(name),
    );
    result.importKind = 'type';
    return result;
  } else if (kind === 'require') {
    return t.variableDeclaration(
      'const',
      [t.variableDeclarator(
        t.identifier(name),
        t.callExpression(
          t.identifier('require'),
          [t.stringLiteral(name)]
        )
      )]
    );
  } else if (kind === 'import') {
    return t.importDeclaration(
      [t.importDefaultSpecifier(t.identifier(name))],
      t.stringLiteral(name)
    );
  } else {
    throw new Error('Kind not specified when creating import node.');
  }
}

function addImportsToBody(t, importNodes, body) {
  let index = 0;
  for (; index < body.length; index++) {
    const bodyNode = body[index];
    if (
      !t.isImportDeclaration(bodyNode)
    ) {
      break;
    }
  }
  const start = body.slice(0, index);
  const end = body.slice(index);
  return [].concat(start, importNodes, end);
}
