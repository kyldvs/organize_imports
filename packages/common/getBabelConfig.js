/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

const DEFAULT_CONFIG = {
  sourceType: 'module',
  plugins: [
    'syntax-async-functions',
    'syntax-flow',
    'syntax-jsx',
    'syntax-object-rest-spread',
    'syntax-trailing-function-commas',
  ],
};

export default function getBabelConfig(config) {
  if (!config) {
    return DEFAULT_CONFIG;
  }
  const result = {};
  const allKeys = [].concat(Object.keys(config), Object.keys(DEFAULT_CONFIG));
  allKeys.forEach(key => {
    if (config[key]) {
      if (Array.isArray(DEFAULT_CONFIG[key])) {
        result[key] = [].concat(DEFAULT_CONFIG[key], config[key]);
      } else {
        result[key] = config[key];
      }
    } else {
      result[key] = DEFAULT_CONFIG[key];
    }
  });
  return result;
};
