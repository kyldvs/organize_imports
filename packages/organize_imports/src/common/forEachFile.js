/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import fs from 'fs';
import path from 'path';

export default function forEachFile(dir, fn) {
  getFiles(dir, dir).forEach(x => fn(x));
}

function getFiles(firstFilePath, originalFilePath) {
  let arr = [];
  const filePaths = fs.readdirSync(originalFilePath);
  filePaths.forEach(fileName => {
    const filePath = path.join(originalFilePath, fileName);
    const stats = fs.statSync(filePath);
    if (stats.isDirectory()) {
      arr = [].concat(arr, getFiles(firstFilePath, filePath));
    } else if (stats.isFile()) {
      arr.push({
        name: fileName,
        relativeName: path.relative(firstFilePath, filePath),
        contents: fs.readFileSync(filePath, 'utf8'),
      });
    }
  });
  return arr;
}
