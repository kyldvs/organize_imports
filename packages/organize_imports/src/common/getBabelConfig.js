/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

const DEFAULT_CONFIG = {
  sourceType: 'module',
  plugins: [
    require('babel-plugin-syntax-async-functions'),
    require('babel-plugin-syntax-flow'),
    require('babel-plugin-syntax-jsx'),
    require('babel-plugin-syntax-object-rest-spread'),
    require('babel-plugin-syntax-trailing-function-commas'),
  ],
};

export default function getBabelConfig(config) {
  if (!config) {
    return DEFAULT_CONFIG;
  }
  const result = {};
  const allKeys = [].concat(Object.keys(config), Object.keys(DEFAULT_CONFIG));
  allKeys.forEach(key => {
    if (config[key]) {
      if (Array.isArray(DEFAULT_CONFIG[key])) {
        result[key] = [].concat(DEFAULT_CONFIG[key], config[key]);
      } else {
        result[key] = config[key];
      }
    } else {
      result[key] = DEFAULT_CONFIG[key];
    }
  });
  return result;
};
