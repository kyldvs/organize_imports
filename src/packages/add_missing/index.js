/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import isFlowDeclarationIdentifier from '../common/isFlowDeclarationIdentifier';
import isRequire from '../common/isRequire';
import globals from 'globals';

const RELEVANT_GLOBALS = new Set(Object.keys({
  ...globals.builtin,
  ...globals.es5,
  ...globals.es6,
  ...globals.browser,
  ...globals.node,
  ...globals.commonjs,
  ...globals.jest,
}));

/**
 * Adds missing imports.
 *
 *   type Options = ?{
 *     importKind?: 'import' | 'require' | 'auto',
 *   };
 *
 */
export default function add_missing(_options) {
  const options = _options || {};
  return function babel_plugin({types: t}) {
    return {
      visitor: {
        Program: {
          exit(path) {
            const maybeUndeclared = new Set(Object.keys(path.scope.globals));

            // Cache of requires to add, based on kind.
            const CACHE = {
              value: new Set(),
              type: new Set(),
            };

            // Determine the kind of import we should add by default.
            let importKind = options.importKind || 'auto';
            const extraVisitors = importKind === 'auto'
              ? {
                ImportDeclaration(path) {
                  if (path.node.importKind === 'value') {
                    importKind = 'import';
                  }
                },
              }
              : {};

            // Find all the identifiers we need to import.
            path.traverse({
              ...extraVisitors,
              Identifier(path) {
                const name = path.node.name;

                if (RELEVANT_GLOBALS.has(name)) {
                  return;
                }

                if (!maybeUndeclared.has(name)) {
                  return;
                }

                let kind = 'value';
                if (t.isGenericTypeAnnotation(path.parentPath.node)) {
                  kind = 'type';
                  if (path.scope.hasBinding(name)) {
                    return;
                  }
                  if (isFlowDeclarationIdentifier(path)) {
                    return;
                  }
                } else if (isFlowPath(path)) {
                  return;
                }

                CACHE[kind].add(name);
              },
            });

            // If it's still auto after the traversal we should use `require`.
            if (importKind === 'auto') {
              importKind = 'require';
            }

            // Add all the requires that we found.
            const added = new Set();
            const importNodes = [];
            CACHE.value.forEach(name => {
              if (!added.has(name)) {
                added.add(name);
                importNodes.push(createImportNode(t, name, importKind));
              }
            });
            CACHE.type.forEach(name => {
              if (!added.has(name)) {
                added.add(name);
                importNodes.push(createImportNode(t, name, 'type'));
              }
            });
            path.node.body = addImportsToBody(t, importNodes, path.node.body);
          },
        },
      },
    };
  };
}

function createImportNode(t, name, kind) {
  if (kind === 'type') {
    const result = t.importDeclaration(
      [t.importDefaultSpecifier(t.identifier(name))],
      t.stringLiteral(name),
    );
    result.importKind = 'type';
    return result;
  } else if (kind === 'require') {
    return t.variableDeclaration(
      'const',
      [t.variableDeclarator(
        t.identifier(name),
        t.callExpression(
          t.identifier('require'),
          [t.stringLiteral(name)]
        )
      )]
    );
  } else if (kind === 'import') {
    return t.importDeclaration(
      [t.importDefaultSpecifier(t.identifier(name))],
      t.stringLiteral(name)
    );
  } else {
    throw new Error('Kind not specified when creating import node.');
  }
}

function addImportsToBody(t, importNodes, body) {
  let index = 0;
  let triggered = false;
  for (; index < body.length; index++) {
    const bodyNode = body[index];
    if (isRequire(bodyNode)) {
      triggered = true;
    } else if (triggered) {
      break;
    }
  }
  if (index === body.length) {
    index = 0;
  }
  const start = body.slice(0, index);
  const end = body.slice(index);
  return [].concat(start, importNodes, end);
}

function isFlowPath(path) {
  if (!path) {
    return false;
  }
  if (path.isFlow()) {
    return true;
  }
  return isFlowPath(path.parentPath);
}
