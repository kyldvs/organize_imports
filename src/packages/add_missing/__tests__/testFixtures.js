/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 */

'use strict';

import add_missing from '../index.js';
import * as babel from 'babel-core';
import forEachFile from '../../common/forEachFile';
import getBabelConfig from '../../common/getBabelConfig';
import path from 'path';

const FIXTURES_DIR = path.join(__dirname, '..', '__fixtures__');

forEachFile(FIXTURES_DIR, ({relativeName, contents}) => {
  test(relativeName, () => {
    expect(run(contents)).toMatchSnapshot();
    expect(run(contents, {importKind: 'auto'})).toMatchSnapshot();
    expect(run(contents, {importKind: 'require'})).toMatchSnapshot();
    expect(run(contents, {importKind: 'import'})).toMatchSnapshot();
  });
});

function run(code, options) {
  return babel.transform(code, getBabelConfig({
    plugins: [
      add_missing(options),
    ],
  })).code;
}
