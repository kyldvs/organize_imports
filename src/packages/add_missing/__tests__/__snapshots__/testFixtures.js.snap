exports[`test auto/detects-imports.js 1`] = `
"
import foo from \'foo\';

import bar from \'bar\';
bar();"
`;

exports[`test auto/detects-imports.js 2`] = `
"
import foo from \'foo\';

import bar from \'bar\';
bar();"
`;

exports[`test auto/detects-imports.js 3`] = `
"
import foo from \'foo\';

const bar = require(\'bar\');

bar();"
`;

exports[`test auto/detects-imports.js 4`] = `
"
import foo from \'foo\';

import bar from \'bar\';
bar();"
`;

exports[`test auto/detects-requires.js 1`] = `
"
const foo = require(\'foo\');

const bar = require(\'bar\');

bar();"
`;

exports[`test auto/detects-requires.js 2`] = `
"
const foo = require(\'foo\');

const bar = require(\'bar\');

bar();"
`;

exports[`test auto/detects-requires.js 3`] = `
"
const foo = require(\'foo\');

const bar = require(\'bar\');

bar();"
`;

exports[`test auto/detects-requires.js 4`] = `
"
const foo = require(\'foo\');

import bar from \'bar\';
bar();"
`;

exports[`test auto/ignores_type_imports.js 1`] = `
"
import type foo from \'foo\';

const bar = require(\'bar\');

bar();"
`;

exports[`test auto/ignores_type_imports.js 2`] = `
"
import type foo from \'foo\';

const bar = require(\'bar\');

bar();"
`;

exports[`test auto/ignores_type_imports.js 3`] = `
"
import type foo from \'foo\';

const bar = require(\'bar\');

bar();"
`;

exports[`test auto/ignores_type_imports.js 4`] = `
"
import type foo from \'foo\';

import bar from \'bar\';
bar();"
`;

exports[`test flow/both_type_and_value_position.js 1`] = `
"const Foo = require(\"Foo\");

const Bar = require(\"Bar\");

// Type position of Foo comes before value position.
function foo(): Foo {
  return new Foo();
}

// Value position of Bar comes before type position.
const theBar = new Bar();
function bar(): Bar {
  return theBar;
}"
`;

exports[`test flow/both_type_and_value_position.js 2`] = `
"const Foo = require(\"Foo\");

const Bar = require(\"Bar\");

// Type position of Foo comes before value position.
function foo(): Foo {
  return new Foo();
}

// Value position of Bar comes before type position.
const theBar = new Bar();
function bar(): Bar {
  return theBar;
}"
`;

exports[`test flow/both_type_and_value_position.js 3`] = `
"const Foo = require(\"Foo\");

const Bar = require(\"Bar\");

// Type position of Foo comes before value position.
function foo(): Foo {
  return new Foo();
}

// Value position of Bar comes before type position.
const theBar = new Bar();
function bar(): Bar {
  return theBar;
}"
`;

exports[`test flow/both_type_and_value_position.js 4`] = `
"import Foo from \"Foo\";
import Bar from \"Bar\";

// Type position of Foo comes before value position.
function foo(): Foo {
  return new Foo();
}

// Value position of Bar comes before type position.
const theBar = new Bar();
function bar(): Bar {
  return theBar;
}"
`;

exports[`test flow/missing_type.js 1`] = `
"import type Foo from \"Foo\";

function foo(x: Foo) {}"
`;

exports[`test flow/missing_type.js 2`] = `
"import type Foo from \"Foo\";

function foo(x: Foo) {}"
`;

exports[`test flow/missing_type.js 3`] = `
"import type Foo from \"Foo\";

function foo(x: Foo) {}"
`;

exports[`test flow/missing_type.js 4`] = `
"import type Foo from \"Foo\";

function foo(x: Foo) {}"
`;

exports[`test flow/nested_types.js 1`] = `
"import type Foo from \"Foo\";
import type Bar from \"Bar\";
import type Buz from \"Buz\";

type Baz = string;

function foo(): ?Foo<Bar<Baz, Buz>> {}"
`;

exports[`test flow/nested_types.js 2`] = `
"import type Foo from \"Foo\";
import type Bar from \"Bar\";
import type Buz from \"Buz\";

type Baz = string;

function foo(): ?Foo<Bar<Baz, Buz>> {}"
`;

exports[`test flow/nested_types.js 3`] = `
"import type Foo from \"Foo\";
import type Bar from \"Bar\";
import type Buz from \"Buz\";

type Baz = string;

function foo(): ?Foo<Bar<Baz, Buz>> {}"
`;

exports[`test flow/nested_types.js 4`] = `
"import type Foo from \"Foo\";
import type Bar from \"Bar\";
import type Buz from \"Buz\";

type Baz = string;

function foo(): ?Foo<Bar<Baz, Buz>> {}"
`;

exports[`test scope-trickery/test1.js 1`] = `
"const bar = require(\"bar\");

function foo() {
  var bar = () => {};
}
bar();

var buz = () => {};
function baz() {
  buz();
}"
`;

exports[`test scope-trickery/test1.js 2`] = `
"const bar = require(\"bar\");

function foo() {
  var bar = () => {};
}
bar();

var buz = () => {};
function baz() {
  buz();
}"
`;

exports[`test scope-trickery/test1.js 3`] = `
"const bar = require(\"bar\");

function foo() {
  var bar = () => {};
}
bar();

var buz = () => {};
function baz() {
  buz();
}"
`;

exports[`test scope-trickery/test1.js 4`] = `
"import bar from \"bar\";

function foo() {
  var bar = () => {};
}
bar();

var buz = () => {};
function baz() {
  buz();
}"
`;

exports[`test simple/empty.js 1`] = `""`;

exports[`test simple/empty.js 2`] = `""`;

exports[`test simple/empty.js 3`] = `""`;

exports[`test simple/empty.js 4`] = `""`;

exports[`test simple/multiple-statements.js 1`] = `
"const foo = require(\"foo\");

foo;
foo;
foo;"
`;

exports[`test simple/multiple-statements.js 2`] = `
"const foo = require(\"foo\");

foo;
foo;
foo;"
`;

exports[`test simple/multiple-statements.js 3`] = `
"const foo = require(\"foo\");

foo;
foo;
foo;"
`;

exports[`test simple/multiple-statements.js 4`] = `
"import foo from \"foo\";
foo;
foo;
foo;"
`;

exports[`test simple/one-statement.js 1`] = `
"const foo = require(\"foo\");

foo;"
`;

exports[`test simple/one-statement.js 2`] = `
"const foo = require(\"foo\");

foo;"
`;

exports[`test simple/one-statement.js 3`] = `
"const foo = require(\"foo\");

foo;"
`;

exports[`test simple/one-statement.js 4`] = `
"import foo from \"foo\";
foo;"
`;

exports[`test simple/skips-first-non-import-statements.js 1`] = `
"
const start = 0; // Date.now() or something.

const foo = require(\'foo\');

const bar = require(\'bar\');

foo();
bar();"
`;

exports[`test simple/skips-first-non-import-statements.js 2`] = `
"
const start = 0; // Date.now() or something.

const foo = require(\'foo\');

const bar = require(\'bar\');

foo();
bar();"
`;

exports[`test simple/skips-first-non-import-statements.js 3`] = `
"
const start = 0; // Date.now() or something.

const foo = require(\'foo\');

const bar = require(\'bar\');

foo();
bar();"
`;

exports[`test simple/skips-first-non-import-statements.js 4`] = `
"
const start = 0; // Date.now() or something.

const foo = require(\'foo\');

import bar from \'bar\';
foo();
bar();"
`;
